src/v/storage/index_state.cc

```diff
@@ -235,6 +240,11 @@ void read_nested(
     if (hdr._version >= index_state::broker_timestamp_version) {
         read_nested(p, st.broker_timestamp, 0U);
     }
+    if (hdr._version >= index_state::first_data_offset_version) {
+        read_nested(p, st.first_data_offset, 0U);
+    } else {
+        st.first_data_offset = tristate<model::offset>{};
```

Comment:
This is only a signal that the state is coming from disk and we don't know anything about the corresponding segment. This is different from the default value which we use for fresh indexes that we're about to add to, which start out in a "has no data batches" state since there are no batches at all, and then updated as we index

Reply:
out of band conversation result: i'm totally sold on tristate being correct.